web/interceptor/access_control.go

```diff
@@ -167,6 +167,13 @@ func (a *AccessControlInterceptor) WrapUnary(next connect.UnaryFunc) connect.Una
 					}
 					return next(ctx, request)
 				}
+				if method == "RebuildSubmissions" || method == "UpdateSubmission" {
+					if !isValidSubmission(a.tokenManager.Database(), req.IDFor("course"), req.IDFor("submission")) {
```

Comment:
Instead of calling these `... req.IDFor("course"), req.IDFor("submission")` here, you can instead pass the `req` to `isValidSubmission` and do:
```go
courseID := req.IDFor("course")
submissionID := req.IDFor("submission")
```
Inside the `isValidSubmission` func.